lib/types.nix

```diff
@@ -395,6 +396,30 @@ rec {
       merge = mergeEqualOption;
     };
 
+    secretFile = let
+      type = mkOptionType {
+        name = "secretFile";
+        description = "quoted path to a secret file";
+        # either the return value of makeWorldReadable or a proper string. We
+        # deliberately refuse paths, because toString x and "${x}" behave
+        # differently. Some modules might handle paths correctly, others will
+        # not, and anyway users can either quote the path or use makeWorldReadable.
+        check = x: ((isCoercibleToString x && (x.canBePublic or false)) || (isString x && !(hasPrefix "/nix/store" x))) && (hasPrefix "/" "${x}");
```

Comment:
Huh indeed the copy happens even when only evaluating builtins like builtins.hasContext... I was very surprised.
```
$ uuidgen > /tmp/secret
$  nix repl "<nixpkgs>"
Welcome to Nix 2.8.1. Type :? for help.

Loading '<nixpkgs>'...
Added 16461 variables.

nix-repl> a = runCommand "foo" {} "cat ${./secret} sdf" 

nix-repl> builtins.hasContext "${a}" 
true

$  rg $(cat /tmp/secret ) /nix/store --glob "*secret"
/nix/store/8gnhfxbv1zzar3pwxlmsxx8lr7876cp4-secret
1:97c31780-96ec-41b1-b8bf-a0bbbe113cb5
```


The check has the following property: a configuration that evaluates does not leak secrets. (well it's not true because you can trick it with `"/${./secret}"` but you get my point). Maybe we could tweak the error message to say that if you see the error message the secret already leaked?